bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java

```diff
@@ -128,9 +139,71 @@ public ThingHandler registerHandler(Thing thing) {
         setHandlerContext(thingHandler);
         registerConfigStatusProvider(thing, thingHandler);
         registerFirmwareUpdateHandler(thing, thingHandler);
+        registerServices(thing, thingHandler);
         return thingHandler;
     }
 
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private void registerServices(Thing thing, ThingHandler thingHandler) {
+        ThingUID thingUID = thing.getUID();
+        for (Class c : thingHandler.getServices()) {
+            Object serviceInstance;
+            try {
+                serviceInstance = c.newInstance();
+
+                if (serviceInstance instanceof ThingHandlerService) {
+                    ThingHandlerService ths = (ThingHandlerService) serviceInstance;
+                    ths.setThingHandler(thingHandler);
+                } else {
+                    logger.warn(
+                            "Should register service={} for thingUID={}, but it does not implement the interface ThingHandlerService.",
+                            c.getCanonicalName(), thingUID);
+                    continue;
+                }
+
+                Dictionary<String, String> properties = new Hashtable<>();
+                properties.put(ThingHandlerService.ACTION_THING_UID, thingUID.getAsString());
+
+                Class[] interfaces = c.getInterfaces();
+                LinkedList<String> serviceNames = new LinkedList<>();
+                if (interfaces != null) {
+                    for (Class i : interfaces) {
+                        serviceNames.add(i.getCanonicalName());
+                    }
+                }
+
+                String[] serviceNamesArray = serviceNames.toArray(new String[serviceNames.size()]);
+
+                ServiceRegistration<?> serviceReg = this.bundleContext.registerService(serviceNamesArray,
+                        serviceInstance, properties);
+
+                if (serviceReg != null) {
+                    Set<ServiceRegistration<?>> serviceRegs = this.thingHandlerServices.get(thingUID);
+                    if (serviceRegs == null) {
+                        HashSet<ServiceRegistration<?>> set = new HashSet<>();
```

Comment:
How would additional services registrations be added here for the same thingUID? This code should be called only once per ThingHandler, right?

Reply:
Got it, fir every service class we add a service registration here.